Delete System.Data.SqlClient package

jkotas opened the pull request:
This package is superseded by Microsoft.Data.SqlClient that lives in https://github.com/dotnet/SqlClient repo.

Fixes https://github.com/dotnet/corefx/issues/40846

jkotas committed:
Delete System.Data.SqlClient package

Fixes https://github.com/dotnet/corefx/issues/40846

jkotas commented on src/libraries/shims/ApiCompatBaseline.netcoreapp.netstandard.txt:
```diff
@@ -387,6 +387,79 @@ CannotMakeMemberNonVirtual : Member 'System.Data.Constraint.ConstraintName.set(S
 CannotMakeMemberNonVirtual : Member 'System.Data.Constraint.Table.get()' is non-virtual in the implementation but is virtual in the contract.
 CannotMakeMemberNonVirtual : Member 'System.Data.Constraint.ToString()' is non-virtual in the implementation but is virtual in the contract.
 CannotMakeMemberNonVirtual : Member 'System.Data.Constraint._DataSet.get()' is non-virtual in the implementation but is virtual in the contract.
+Compat issues with assembly System.Data.SqlClient:
```
I am not sure whether this is the best way to suppress these.

stephentoub replied:
What contract is complaining about these?  The System.Data.dll shim?  What implications does that have for libraries built against the netfx surface area trying to run on core?

ericstj replied:
We need those shims to continue to work.  You can restore an previous version of the package to produce the correct shims.

jkotas replied:
> You can restore an previous version of the package to produce the correct shims.

I have added it to https://github.com/dotnet/runtime/pull/2275/files#diff-d99c84c0369652bea55f739641b3ef74R40 . Is this the right place to restore the package?

stephentoub commented on src/coreclr/tests/scripts/run-corefx-tests.bat:
```diff
@@ -27,7 +27,7 @@ echo control, but is easy to implement. This file should be a list of assembly n
 echo.
 echo     System.Console.Tests
-echo     System.Data.SqlClient.Tests
+echo     System.Data.Common.Tests
```
I can't comment on it, but there's a `src/libraries/Native/pkg/_.pdb` in the diff file list, showing as "No changes".  Do you know what's going on there?

jkotas replied:
The whole src/libraries/Native/pkg directory is gone. `_pdb` was an empty file. I guess it is what makes gihub to show the file this way.

stephentoub commented:
What about the references in
https://github.com/dotnet/runtime/blob/master/src/libraries/pkg/descriptions.json
https://github.com/dotnet/runtime/blob/master/src/libraries/pkg/Microsoft.Private.PackageBaseline/packageIndex.json
https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/src/mono/netcore/CoreFX.issues_windows.rsp#L30
https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/src/mono/mono/metadata/assembly.c#L163
https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/.config/CredScanSuppressions.json#L20-L25
etc.?

jkotas commented:
> https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/src/libraries/pkg/descriptions.json#L550-L568
> https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/src/libraries/pkg/Microsoft.Private.PackageBaseline/packageIndex.json#L1416-L1430

These list have references to all packages that were ever built in CoreFX, even many dead ones. I have kept it that way for consistency. @ericstj Let me know if SqlClient should be cleaned up from these list.

https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/src/mono/netcore/CoreFX.issues_windows.rsp#L30

Fixed.

> https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/src/mono/mono/metadata/assembly.c#L163

This list is for legacy Mono. Nothing is changing for legacy Mono with this change. These sources live in dotnet/runtime as side-effect of bulk source sharing with legacy Mono.

> https://github.com/dotnet/runtime/blob/b8942c37f25460bb160d0ff2fde9795d56b9bf01/.config/CredScanSuppressions.json#L20-L25

Fixed.

ericstj commented:
> packageIndex.json

The PackageIndex.json shouldn't be changed. We need this in order to emit the right package dependencies for libraries which depend on SqlClient.  It looks like OleDb still depends on SqlClient. https://github.com/dotnet/runtime/blob/3b9abae5aae537258611b5750dc7e2963cfc73ed/src/libraries/System.Data.OleDb/src/System.Data.OleDb.csproj

ericstj commented on src/libraries/restore/binplacePackages/binplacePackages.depproj:
```diff
@@ -37,6 +37,7 @@
   
   <ItemGroup>
     <PackageReference Include="Microsoft.Bcl.AsyncInterfaces" Version="1.0.0" Condition="'$(TargetsNetStandard)' != 'true' or '$(NETStandardVersion)' &gt;= 2.0" />
+    <PackageReference Include="System.Data.SqlClient" Version="4.8.0" Condition="'$(TargetFramework)' == '$(NetCoreAppCurrent)'" />
```
This is the right place, but I believe you'll also need it for netstandard https://github.com/dotnet/runtime/blob/23c54f183c9faf2049223a44197c6b8b282b90c4/src/libraries/System.Data.OleDb/src/Configurations.props#L5.

Wraith2 commented:
> If that lib does indeed need SqlClient, then we should also restore the old package to replace the live-built ref.

It's referenced but only one file imports the namespace and nothing from it is used. It's trivial to remove the dependency.

jkotas committed:
Delete unnecessary SqlClient references

ericstj commented:
Much better to remove it completely if it isn’t needed.

We will still need something like the restore step to get the System.Data shim correct.

jkotas commented:
> Much better to remove it completely if it isn’t needed.
> We will still need something like the restore step to get the System.Data shim correct.

I have done both these. Thanks!

The last problem that I have is how to drop System.Data.SqlClient.dll into the testhost. System.Data.Common tests depend on it. What is the best way to drop System.Data.SqlClient.dll into the testhost?

Wraith2 commented:
All the code references in the tests seem to be from `DbProviderFactoriesTests` and could be replaced with another provider as far as I can see.

jkotas committed:
Unify System.Data.SqlClient package reference

jkotas committed:
Add System.Data facade test

ViktorHofer commented:
Doesn't binplacing System.Data.SqlClient from a package into the runtime/testhost directory mean that during sourcebuild we need that package as a pre-built?

Wraith2 commented:
If you only need the sqlclient library for the System.Data tests you could just replace the provider tests with OleDB or ODBC as the provider and then you don't need sqlclient or it's tests for this build do you?

jkotas commented:
I need System.Data.SqlClient references for the source build, and I would like to keep it for running tests as well (e.g. the new test that I am adding in this PR). I do not think it should be a problem for source build:
- References are handled via ilasm from .il for source build
- Tests are not a problem for source build.

ericstj commented on src/libraries/shims/manual/System.Data.csproj:
```diff
@@ -2,4 +2,8 @@
\ No newline at end of file
+  <ItemGroup>
+    <!-- System.Data.SqlClient is not live built in anymore -->
+    <ReferencePath Include="$(ArtifactsBinDir)ref\$(NetCoreAppCurrent)\System.Data.SqlClient.dll" />
```
Should we copy to runtime as well?   That would avoid this workaround as well as other PackageReferences.

ViktorHofer replied:
The runtime/testhost directory currently contains only shared framework assets. I'm a bit worried that adding System.Data.SqlClient there conflicts with the principle of testing the live built shared framework. I admit, we are also putting OOB assemblies there but that's just for convenience atm.

jkotas replied:
> we are also putting OOB assemblies there but that's just for convenience atm.

Right, the runtime/testhost contains a pile of stuff that should not be there. So this is a question of whether to be consistent with the convenience we have there today, or whether to do work towards cleaning it up. I went back and forth on this. Neither of these options are 100% pretty.

ViktorHofer replied:
> or whether to do work towards cleaning it up

I started cleaning it up when I refactored the testing logic in corefx and removed all testing assets. Currently there shouldn't be any cruft besides the OOB assets, CoreRun which was brought back for profiling/performance testing.

I planned to remove the OOB assets from there as well when we did the work to make our test libraries consumable by Mono. This isn't a high prio anymore but I still tend towards making the OOB libraries app local in the test output folders instead of having them in the shared framework testhost.

jkotas replied:
> I planned to remove the OOB assets

I agree that this is cleaner and it is what the current shape of the PR is trying to do: I have added System.Data.SqlClient package references for the two tests that need them. But that it is colliding with other other tricks that the build plays and causing Mono and allConfigurations legs to fail.

jkotas replied:
> Should we copy to runtime as well?

Trying this alternative now.

ViktorHofer replied:
OK. Do you understand the Mono failures? https://dnceng.visualstudio.com/public/_build/results?buildId=506541&view=ms.vss-test-web.build-test-results-tab&runId=16074090&resultId=110670&paneView=debug

jkotas replied:
I believe that some files are getting overwritten during the build with a wrong copy.

ViktorHofer replied:
> Should we copy to runtime as well?

I wasn't following, why do we need to copy this into the runtime/testhost folder?

ericstj commented:
> Doesn't binplacing System.Data.SqlClient from a package into the runtime/testhost directory mean that during sourcebuild we need that package as a pre-built?

Depends on what part of the build needs it.  Today, we already have a prebuilt requirement for shims which is the only part of the product build that requires this.  We do a two-pass build, where first pass can consume the reference, second pass can use the generated source.  I think we can piggy-back on that.

Tests are already are allowed to have prebuilts in source build.

jkotas committed:
Add System.Data.SqlClient references
- Add System.Data.SqlClient package references for tests that need it
- Use System.Data.SqlClient reference assembly to build the System.Data shim
- Disable ILLinker for the shims since it does not work well with incremental builds and it is not necessary anyway

Try to use package references in tests

jkotas committed:
Disable tests on Mono

safern commented:
> The last problem that I have is how to drop System.Data.SqlClient.dll into the testhost.

We already restore test projects, why not just add a `PackageReference` in the test project to the given package?

jkotas commented:
> why not just add a PackageReference in the test project to the given package?

This is what I have done - the last two commits are the change required to flip from including it in the runtime to making the PackageReference work). It is hitting a bug in Mono assembly loader. I have disabled the affected test and will file an issue on it if the rest looks good.

This is passing the CI now. The change should be ready to review!

ViktorHofer commented on src/libraries/shims/manual/System.Data.csproj:
```diff
@@ -2,4 +2,9 @@
+  <ItemGroup>
+    <!-- System.Data.SqlClient is not live built anymore, make sure to use reference assembly for it -->
+    <ReferencePath Remove="$(RefPath)\System.Data.SqlClient.dll" />
+    <ReferencePath Include="$(ArtifactsBinDir)ref\$(NetCoreAppCurrent)\System.Data.SqlClient.dll" />
```
Nit: why not use [RefPath](https://github.com/dotnet/runtime/blob/95b361d7b9551e59f679ecd75cfc1480ee4e9c59/src/libraries/Directory.Build.props#L252) here as well?

jkotas replied:
Because of it is re-defined to mean something else [here](https://github.com/dotnet/runtime/blob/master/src/libraries/shims/manual/Directory.Build.props#L15).

ViktorHofer commented on src/libraries/System.Runtime.Serialization.Formatters/tests/System.Runtime.Serialization.Formatters.Tests.csproj:
```diff
@@ -32,4 +32,7 @@
     <Reference Include="System.Text.Json" />
   </ItemGroup>
+  <ItemGroup>
+    <PackageReference Include="System.Data.SqlClient" Version="4.8.0" />
```
Can we define that version in Versions.props instead and reuse it in both test projects?

jkotas replied:
Done

ViktorHofer reviewed (approved):
Thanks, LGTM. I would like to know why we need to copy into the runtime/testhost before merging though.

jkotas committed:
Define System.Data.SqlClient version in Versions.props

jkotas commented:
>  why we need to copy into the runtime/testhost before merging though.

We do not copy it in my latest version of the change. I have switched the tests that need it to have local package references and removed copying of `System.Data.SqlClient.dll` into runtime/testhost.

For reference, here is the list of places where `System.Data.SqlClient.dll` is placed under `artifacts\bin` after running `build.cmd`. All of them are refs.
```
\runtime\artifacts\bin\binplacePackages\netcoreapp5.0-Release\System.Data.SqlClient.dll
\runtime\artifacts\bin\binplacePackages\netstandard2.0-Release\System.Data.SqlClient.dll
\runtime\artifacts\bin\netstandard\netstandard2.0-Release\System.Data.SqlClient.dll
\runtime\artifacts\bin\ref\netcoreapp5.0\System.Data.SqlClient.dll
\runtime\artifacts\bin\ref\netstandard2.0\System.Data.SqlClient.dll
```

ViktorHofer commented:
Thanks, I believe I saw the PackageReference in runtime.depproj but maybe that was an old commit. Thanks for not writing into the runtime/testhost folder 👍 +💯

jkotas commented:
The test failures are known issues:
- Failed to acquire dotnet install script
- Docker container recycling problem

safern commented:
@jkotas it seems like this broke the rolling build tests on Linux_musl_arm64 (Alpine).

Binary formatter tests are failing: https://helix.dot.net/api/2019-06-17/jobs/ed130407-35ad-40cc-a9ef-24b9767b2a77/workitems/System.Runtime.Serialization.Formatters.Tests/console

The Binary Data that is failing seems to be System.Data objects. L581 of https://github.com/dotnet/runtime/blob/master/src/libraries/System.Runtime.Serialization.Formatters/tests/BinaryFormatterTestData.cs#L581 (GH won't show it since it is too big).

Also, System.Data.Common: https://helix.dot.net/api/2019-06-17/jobs/cb568d05-e0c9-4b5d-a292-1ac6dd5847a9/workitems/System.Data.Common.Tests/console

https://dev.azure.com/dnceng/public/_build/results?buildId=509595&view=ms.vss-test-web.build-test-results-tab

It could be related to the asset in the package reference having a bug or not picking the right one for Linux_musl_arm64 but not sure.

jkotas commented:
> It could be related to the asset in the package reference having a bug or not picking the right one for Linux_musl_arm64 but not sure.

Right, the problem is that the build for musl picks up `lib\netcoreapp2.1\System.Data.SqlClient.dll` from the https://www.nuget.org/packages/System.Data.SqlClient/ . This .dll is a stub implementation that throws PlatformNotSupportedExceptions in most methods.

Instead, it needs to pick up `runtimes\unix\lib\netcoreapp2.1\System.Data.SqlClient.dll` . 

I have verified that a simple console app published for must gets the right one. So this has to do with some customization of package restore that we got in the build. Do you happen to know where that may happen?

A simple way to fix this is to switch back to copying System.Data.SqlClient.dll to testhost. I will do that.

jkotas commented:
#31850
